### PR TITLE
Reformatted stable_pow

### LINKED ISSUE #2228: stable_pow

### PR DESCRIPTION
**Pull request by Vandith2**
Fixes #2228

### CODE REVIEW

**Comment by Vandith2** [comment]
@RitujaPawas I have edited the docstring to follow guidelines

**Comment by RitujaPawas** [comment]
HI @Vandith2 Changes look good. I will take a final look and let you know!

**Comment by Vandith2** [comment]
@RitujaPawas are there any changes to be made?

**Comment by RitujaPawas** [comment]
> @RitujaPawas are there any changes to be made?
 Please add instance methods examples!

**Comment by Vandith2** [comment]
@RitujaPawas I have added the instance method examples, is that fine or do I need to change the first example as well?

**Comment by RitujaPawas** [comment]
Also, make sure the docstrings tests and numpy-docstring-style checks are passing for your function! let me know if you need any help!

**Comment by Vandith2** [comment]
Could you please help me with the docstring tests? I'm not sure why it's failing.

**Comment by RitujaPawas** [comment]
> Could you please help me with the docstring tests? I'm not sure why it's failing.

Sure, I will take a look and let you know!

**Comment by RitujaPawas** [comment]
Hi @Vandith2 
1-please remove the unnecessary files added to your changes?
2-please update your fork to the latest ivy

**Comment by VedPatwardhan** [comment]
Thanks for your reviews @RitujaPawas. Sorry for jumping in @Vandith2, but I'll be reviewing this PR going forward to ease the load for @RitujaPawas. Please make the changes as requested by @RitujaPawas and re-request my review once you're done. Thanks 🙂

**Comment by Vandith2** [comment]
@VedPatwardhan I have made the changes requested.

File: ivy/array/general.py [hunk 1]
```diff
@@ -472,6 +472,51 @@ def stable_divide(
...
+
+        Examples
+        --------
+        >>> x = ivy.asarray([2, 4, 5])
+        >>> y = x.stable_pow(2)
+        >>> print(y)
+        ivy.array([ 4, 16, 25])
+
+        >>> x = ivy.asarray([2, 4, 5])
+        >>> y = ivy.Array.stable_pow(x, 4, min_base=1)
```
[the remark this answers is not in the document]
Comment (VedPatwardhan) [reply ivy/array/general.py:507]: This applied to the other instance method examples as well.

File: ivy/array/general.py [hunk 2]
```diff
@@ -472,6 +472,51 @@ def stable_divide(
...
+
+        Returns
+        -------
+        ret
+            Numpy array containing the elements of base raised to the power of the
+            corresponding element of exponent.
+
+        Examples
+        --------
+        >>> x = ivy.asarray([2, 4, 5])
```
[the remark this answers is not in the document]
Comment (VedPatwardhan) [reply ivy/array/general.py:501]: This applies to all other docstring examples you've added.

File: ivy/container/general.py [hunk 3]
```diff
@@ -2134,3 +2135,199 @@ def clip_matrix_norm(
...
+        >>> x = ivy.Container(a=3, b=10)
+        >>> y = ivy.asarray([4, 5])
+        >>> z = ivy.Container.static_stable_pow(x, y)
+        >>> print(z)
+        {
+            a: array([81.00108001, 243.00405003]),
+            b: array([10000.04000006, 100000.500001])
+        }
+
+        >>> x = ivy.Container(a=3, b=10)
```
Comment (VedPatwardhan) [inline ivy/container/general.py:2199]: This example should be removed as we should not having native arrays in the docstring examples.

File: ivy/container/general.py [hunk 4]
```diff
@@ -2134,3 +2135,199 @@ def clip_matrix_norm(
...
+        >>> y = ivy.Container(a=ivy.asarray([0.5, 2.5]), b=ivy.asarray([3.5, 1.5]))
+        >>> z = ivy.Container.static_stable_pow(x, y)
+        >>> print(z)
+        {
+            a: ivy.array([2.65, 5.66]),
+            b: ivy.array([46.8, 8.])
+        }
+
+        >>> x = ivy.Container(a=ivy.asarray([1.5, 2.], [3., 4.]),\
+        ... b=ivy.asarray([5., 6.], [7., 8.]))
```
Comment (VedPatwardhan) [inline ivy/container/general.py:2218]: Please remove the `...` and adjust the indentation such that `b` comes right below `a`.

File: ivy/functional/ivy/general.py [hunk 5]
```diff
@@ -2290,23 +2292,90 @@ def stable_divide(
...
+    >>> print(z)
+    ivy.array([16., 64., 36.])
+
+    >>> x = ivy.asarray([2.0, 4.0, 81.0])
+    >>> y = ivy.asarray([2, 0.5, 3])
+    >>> z = ivy.asarray([0.1, 0.2, 0.3])
+    >>> print(ivy.stable_pow(x, y, min_base = z))
+    ivy.array([4.41e+00, 2.05e+00, 5.37e+05])
+
+    With :code:`ivy.Container` input:
```
Comment (VedPatwardhan) [inline ivy/functional/ivy/general.py:2360]: Please separate out the 2 examples under this header as `With one ivy.Container input` and `With multiple ivy.Container inputs`/

File: ivy/functional/ivy/general.py [hunk 6]
```diff
@@ -2290,23 +2292,90 @@ def stable_divide(
...
 
     Returns
     -------
     ret
-        The new item following the numerically stable division.
+        The new item following the numerically stable exponentiation.
+
+    Examples
+    --------
+    With :code:`int` input:
```
Comment (VedPatwardhan) [inline ivy/functional/ivy/general.py:2319]: Please club the 3 headers for `int`, `float` and `complex` under one header `With Number input`

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hi @Vandith2, thanks for making the changes. I have suggested a few more changes near the lines of code concerned. Please also revert the changes made to the `.idea/misc.xml` file in your PR such that it doesn't appear under the `Files Changed` tab. Secondly, there are a few changes to be made to get the docstring tests to pass for your function, this is the log of the tests regarding the same,
```
output for  stable_pow  on run:  {a:ivy.array([1000.,3380.]),b:ivy.array([8000.,15600.])},{a:DeviceArray([81.00108001,243.00405003],dtype=float64,weak_type=True),b:DeviceArray([10000.04000006,100000.500001],dtype=float64,weak_type=True)},{a:array([140.29821987,19683.59049787]),b:array([3.16229189e+04,1.00000900e+09])},{a:ivy.array([2.65,5.66]),b:ivy.array([46.8,8.])},invalidsyntax(<string>,line1)jaxx=ivy.Container(a=ivy.asarray([1.5,2.],[3.,4.]),...b=ivy.asarray([5.,6.],[7.,8.])),{a:ivy.array([3.,32.]),b:ivy.array([280.,14.7])} 
output in docs : {a:ivy.array([1000.,3380.]),b:ivy.array([8000.,15600.])},{a:array([81.00108001,243.00405003]),b:array([10000.04000006,100000.500001])},{a:array([140.29821987,19683.59049787]),b:array([3.16229189e+04,1.00000900e+09])},{a:ivy.array([2.65,5.66]),b:ivy.array([46.8,8.])},{a:ivy.array([1.87,32.]),b:ivy.array([907.,22.6])} 
 DeviceArray81.00108001  !=  array81.00108001 
```
These changes are related to the precision of the output, please run the examples locally and copy those results into the docstring to avoid this from happening.
One more suggestion, the changes requested don't just apply to the example where they're mentioned, but the same changes should be applied to the other examples too which have that problem.
Feel free to re-request my review through the right pane once you're done making the changes. Thanks 🙂

**Comment by VedPatwardhan** [comment]
Hi @Vandith2, please message in the chats if you're still working on this PR, because it may get closed automatically in a few days due to inactivity. Thanks 🙂

**Comment by Vandith2** [comment]
Hi, I'm still working on it.

**Comment by Vandith2** [comment]
Hi, I'm still working on it.

**Comment by VedPatwardhan** [comment]
Hi @Vandith2, please message in the chats if you're still working on this PR, because it may get closed automatically in a few days due to inactivity. Thanks 🙂

**Comment by Vandith2** [comment]
Hi, sorry for the late reply. I've been busy the past few days but I am working on the change right now.

File: ivy/functional/ivy/general.py (hunk 5 above)
Comment (Vandith2) [reply ivy/functional/ivy/general.py:2360]: Should I change that for array examples as well?

Comment (VedPatwardhan) [reply ivy/functional/ivy/general.py:2360]: No, this structure is meant only for Container examples, because we want to demonstrate that the function is nestable, i.e. the function works when we pass an `ivy.Container` in place of any or all the arguments.

Comment (Vandith2) [reply ivy/functional/ivy/general.py:2360]: Got it, thank you

**Comment by Vandith2** [comment]
Hi, for some reason, I'm not able to run stable_pow as an array or container instance to test my instance examples.
It gives me this error:
`Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/ivy/ivy/container/wrapping.py", line 38, in new_function
    data_idx = function.array_spec[0]
IndexError: list index out of range`

File: ivy/array/general.py [hunk 7]
```diff
@@ -657,9 +657,73 @@ def stable_divide(
             a numpy array containing the elements of numerator divided by
             the corresponding element of denominator
 
+        Examples
```
Comment (VedPatwardhan) [inline ivy/array/general.py:660]: Please revert these changes as it's not related to the function being reformatted through your task.

File: ivy/array/general.py [hunk 8]
```diff
@@ -657,9 +657,73 @@ def stable_divide(
...
+        >>> y = ivy.array([[1., 2., 3.], [2., 3., 4.]])
+        >>> z = x.stable_divide(y)
+        >>> print(z)
+        ivy.array([[4.  , 2.5 , 2.  ],
+               [3.5 , 2.67, 2.25]])
+
         """
         return ivy.stable_divide(self, denominator, min_denominator=min_denominator)
 
+    def stable_pow(
```
Comment (VedPatwardhan) [inline ivy/array/general.py:682]: There seems to be a duplicate `stable_pow` function in the same file

File: ivy/functional/ivy/general.py [hunk 9]
```diff
@@ -2467,26 +2469,81 @@ def stable_divide(
 @inputs_to_native_arrays
 @handle_nestable
 def stable_pow(
-    base: Union[Number, ivy.Array, ivy.NativeArray],
-    exponent: Union[Number, ivy.Array, ivy.NativeArray],
-    min_base: float = None,
-) -> Any:
+    base: Union[Number, ivy.Array, ivy.NativeArray, ivy.Container],
```
Comment (VedPatwardhan) [inline ivy/functional/ivy/general.py:2472]: Please don't update the typehints here, also you don't need to add `ivy.Container` to the typehints because every nestable function supports an `ivy.Container` in any or all the arguments, which need not be specified in the functional version.

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hi @Vandith2, I have suggested a few more changes to your PR. The issue you're facing could be related to the redefinition of the `stable_pow` function in `array/general.py`, please look into it. Also, feel free to re-request my review once you're done. Thanks 🙂

**Comment by Vandith2** [comment]
Hi, for some reason I am unable to resolve the conflict. Every time I edit it and try to merge, it says conflict resolved but if I refresh it still says there are conflicts.

**Comment by VedPatwardhan** [comment]
Hi @Vandith2, I have resolved the conflict now, please re-request my review once you're done with the changes. Thanks 🙂

**Comment by Vandith2** [comment]
Hi, thanks.
I had already made the changes requested.

File: ivy/array/general.py [hunk 10]
```diff
@@ -641,7 +641,6 @@ def stable_divide(
         ivy.Array instance method variant of ivy.stable_divide. This method simply wraps
         the function, and so the docstring for ivy.stable_divide also applies to this
         method with minimal changes.
-
```
Comment (VedPatwardhan) [inline ivy/array/general.py:644]: Please revert this change as it isn't related to your function.

File: ivy/array/general.py [hunk 11]
```diff
@@ -650,16 +649,59 @@ def stable_divide(
             denominator for division.
         min_denominator
             the minimum denominator to use, use global ivy._MIN_DENOMINATOR by default.
-
```
Comment (VedPatwardhan) [inline ivy/array/general.py:653]: Please revert this change as it isn't related to your function.

File: ivy/array/general.py [hunk 12]
```diff
@@ -650,16 +649,59 @@ def stable_divide(
...
         Returns
         -------
         ret
             a numpy array containing the elements of numerator divided by
             the corresponding element of denominator
-
```
Comment (VedPatwardhan) [inline ivy/array/general.py:659]: Please revert this change as it isn't related to your function.

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hi @Vandith2, thanks for making the changes. I have requested a few more changes near the lines of code concerned. Feel free to re-request my review once you're done with the changes and we'll check if the tests pass.
Let me know if you have any questions. Thanks 🙂

**Comment by Vandith2** [comment]
Hi, I have made the changes requested.

File: ivy/array/general.py [hunk 13]
```diff
@@ -36,9 +36,9 @@ def is_native_array(
 
     def is_ivy_array(self: ivy.Array, /, *, exclusive: Optional[bool] = False) -> bool:
         """
-        ivy.Array instance method variant of ivy.is_ivy_array. This method simply 
-        wraps the function, and so the docstring for ivy.is_ivy_array also 
-        applies to this method with minimal changes.
+        ivy.Array instance method variant of ivy.is_ivy_array. This method simply
```
Comment (VedPatwardhan) [inline ivy/array/general.py:39]: Please revert this update as it's not related to the function you're reformatting.

File: ivy/array/general.py [hunk 14]
```diff
@@ -560,6 +560,51 @@ def stable_divide(
         """
         return ivy.stable_divide(self, denominator, min_denominator=min_denominator)
 
+        def stable_pow(
+        self: ivy.Array,
```
Comment (VedPatwardhan) [inline ivy/array/general.py:564]: Please fix the indentation here

File: ivy/array/general.py [hunk 15]
```diff
@@ -817,32 +862,24 @@ def default(
         """
         return ivy.default(self, default_val, catch_exceptions, rev, with_callable)
 
-    def stable_pow(
-        self: ivy.Array,
-        exponent: Union[Number, ivy.Array, ivy.NativeArray],
-        min_base: float = None,
-    ) -> Union[ivy.Array, ivy.NativeArray]:
+    def supports_inplace(self: ivy.Array) -> bool:
```
Comment (VedPatwardhan) [inline ivy/array/general.py:865]: Please revert this change as it's not related to the function you're reformatting

File: ivy/container/general.py [hunk 16]
```diff
@@ -6,6 +6,7 @@
 from ivy.container.base import ContainerBase
 import ivy
 
+
```
Comment (VedPatwardhan) [inline ivy/container/general.py:9]: Please remove this additional line

File: ivy/container/general.py [hunk 17]
```diff
@@ -3532,3 +3533,185 @@ def supports_inplace_updates(
             prune_unapplied=prune_unapplied,
             map_sequences=map_sequences,
         )
+        
+    @staticmethod
+    def static_stable_pow(
```
Comment (VedPatwardhan) [inline ivy/container/general.py:3538]: There is another function with the same name in this file, please make sure that there's only one function named `static_stable_pow`

File: ivy/container/general.py [hunk 18]
```diff
@@ -3532,3 +3533,185 @@ def supports_inplace_updates(
...
+            base,
+            exponent,
+            min_base,
+            key_chains=key_chains,
+            to_apply=to_apply,
+            prune_unapplied=prune_unapplied,
+            map_sequences=map_sequences,
+        )
+
+    def stable_pow(
```
Comment (VedPatwardhan) [inline ivy/container/general.py:3630]: There is another function with the same name in this file, please make sure that there's only one function named `stable_pow`

File: ivy/container/general.py [hunk 19]
```diff
@@ -3532,3 +3533,185 @@ def supports_inplace_updates(
...
+
+        """
+        return self.static_stable_pow(
+            self,
+            exponent,
+            min_base=min_base,
+            to_apply=to_apply,
+            prune_unapplied=prune_unapplied,
+            map_sequences=map_sequences,
+        )
```
Comment (VedPatwardhan) [inline ivy/container/general.py:3717]: Please add a blank line at the end of this file

File: ivy/functional/ivy/general.py [hunk 20]
```diff
@@ -2672,6 +2727,290 @@ def inplace_increment(
     return current_backend(x).inplace_increment(x, val)
 
 
+@to_native_arrays_and_back
```
Comment (VedPatwardhan) [inline ivy/functional/ivy/general.py:2730]: Please revert this change as it's not related to the function you're reformatting

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hi @Vandith2, thanks for making the changes, I have requested a few more changes regarding some issues which may have been caused due to incorrect merge conflict resolution, please make those changes. Also, another suggestion, please re-request my review through the right pane on this page once you're done with the changes, because it gets difficult for me to understand if you're done with the changes otherwise.

Let me know if you have any questions. Thanks 🙂

File: ivy/container/general.py [hunk 21]
```diff
@@ -3609,3 +3501,185 @@ def supports_inplace_updates(
...
+            b: array([[907.4927, 22.627417],
+                      [2187., 31.622776]], dtype=float32)
+        }
+
+        """
+        return ContainerBase.multi_map_in_static_method(
+            "stable_pow",
+            base,
+            exponent,
+            min_base,
```
Comment (VedPatwardhan) [inline ivy/container/general.py:3591]: `min_base` should be passed as a keyword argument here

File: ivy/functional/ivy/general.py [hunk 22]
```diff
@@ -2226,16 +2226,71 @@ def stable_pow(
...
+
+    >>> x = ivy.Container(a = ivy.array([3., 4.]), b = ivy.array([9., 6.]))
+    >>> y = ivy.Container(a = ivy.array([2, 3]), b = ivy.array([0.5, 4]))
+    >>> z = ivy.stable_pow(x, y)
+    >>> print(z)
+    {
+        a: ivy.array([9., 64.]),
+        b: ivy.array([3., 1300.])
+    }
+
```
Comment (VedPatwardhan) [inline ivy/functional/ivy/general.py:2293]: There are 2 additional blank lines here which should be removed.

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hi @Vandith2, just suggested some final changes to the PR before we merge it. Thanks 🙂

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hi @Vandith2, thanks for making the changes. There is one more issue I noticed in the `stable_pow` function as it was having the wrong decorator. I have pushed a fix to master, please pull the code from master so that it gets reflected in your PR too. Basically, in the container examples, the outputs of the examples should be `ivy.array` when you run them after those changes are pulled, so you'll have to update the outputs in those examples. Also, I'd suggest testing all other examples in the array instance, container instance and static methods and the functional version to see if there are any other changes to be made to the outputs of those examples.
Feel free to re-request my review once you're done with the changes. Thanks 🙂

**Comment by Vandith2** [comment]
Hi, I already ran all the examples and pasted the outputs exactly as I got them.

File: ivy/container/general.py [hunk 23]
```diff
@@ -3609,3 +3501,185 @@ def supports_inplace_updates(
...
+            stable exponentiation.
+
+        Examples
+        --------
+        >>> x = ivy.Container(a=ivy.array([10., 15.]), b=ivy.array([20., 25.]))
+        >>> y = ivy.Container.static_stable_pow(x, 3)
+        >>> print(y)
+        {
+            a: ivy.array([1000.00287, 3375.0063], dtype=float32),
+            b: array([8000.011, 15625.018], dtype=float32)
```
Comment (VedPatwardhan) [inline ivy/container/general.py:3553]: The result of this example is incorrect, every ivy function returns an `ivy.Array` when it receives an `ivy.Array`.

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hi @Vandith2, thanks for making the changes. The result of an example seems to be incorrect, as I mentioned in a previous review I had pushed a fix for that function to master, so maybe pull the latest code from master into your branch and check again. Thanks 🙂

**Comment by VedPatwardhan** [comment]
Hi @Vandith2, please message in the chats if you're still working on this PR, because it may get closed automatically in a few days due to inactivity. Thanks 🙂

**Comment by Vandith2** [comment]
Hi, yeah I am working on it.

**Comment by VedPatwardhan** [comment]
Hi @Vandith2, please message in the chats if you're still working on this PR, because it may get closed automatically in a few days due to inactivity. Thanks 🙂

**Comment by Vandith2** [comment]
Hi, sorry I was busy but I am still working on it.

**Comment by VedPatwardhan** [comment]
Hi @Vandith2, please message in the chats if you're still working on this PR, because it may get closed automatically in a few days due to inactivity. Thanks 🙂

**Comment by Vandith2** [comment]
Hi, sorry for the late reply, I have been busy with course work and I recently had to reset my laptop, so I am still setting up all the software I had installed but I am still working on this.

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by ivy-seed** [comment]
This PR has been closed because it has been marked as stale for more than 7 days with no activity.

**Comment by ivy-seed** [comment]
This PR has been closed because it has been marked as stale for more than 7 days with no activity.
